constraint-solver/src/solver/base.rs

```diff
@@ -263,23 +264,21 @@ where
 {
     /// Performs boolean extraction on `constr`, i.e. tries to turn quadratic constraints into affine constraints
     /// by introducing new boolean variables.
-    /// This function will always return the original constraint as well as any extracted constraints.
-    fn extract_boolean(
+    /// Only returns the extracted constraint, not the original.
+    fn try_extract_boolean(
         &mut self,
-        constr: GroupedExpression<T, V>,
-    ) -> impl Iterator<Item = GroupedExpression<T, V>> {
+        constr: &GroupedExpression<T, V>,
+    ) -> Option<GroupedExpression<T, V>> {
         let mut new_boolean_var = None;
         let extracted = self
             .boolean_extractor
-            .try_extract_boolean(&constr, &mut || {
+            .try_extract_boolean(constr, &mut || {
                 let v = self.var_dispenser.next_boolean();
                 new_boolean_var = Some(v.clone());
                 v
-            });
-        if let Some(v) = new_boolean_var {
-            self.add_range_constraint(&v, RangeConstraint::from_mask(1))
-        }
-        std::iter::once(constr).chain(extracted)
+            })?;
+        self.add_range_constraint(&new_boolean_var.unwrap(), RangeConstraint::from_mask(1));
```

Comment:
```
self.assert_is_boolean(&new_boolean_var.unwrap())
```

maybe we could have a method like this

Reply:
I think this is the only place this would be used.